File: ItemManager.cs
using System;
using System.Collections.Generic;

/*
 * Copyright (c) 2016-2017, Adam <Adam@sigterm.info>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
namespace OSRSCache
{
	using ItemDefinition = OSRSCache.definitions.ItemDefinition;
	using ItemExporter = OSRSCache.definitions.exporters.ItemExporter;
	using ItemLoader = OSRSCache.definitions.loaders.ItemLoader;
	using ItemProvider = OSRSCache.definitions.providers.ItemProvider;
	using Archive = OSRSCache.fs.Archive;
	using ArchiveFiles = OSRSCache.fs.ArchiveFiles;
	using FSFile = OSRSCache.fs.FSFile;
	using Index = OSRSCache.fs.Index;
	using Storage = OSRSCache.fs.Storage;
	using Store = OSRSCache.fs.Store;
	// using IDClass = OSRSCache.util.IDClass;

	public class ItemManager : ItemProvider
	{
		private readonly Store store;
		private readonly IDictionary<int, ItemDefinition> items = new Dictionary<int, ItemDefinition>();

		public ItemManager(Store store)
		{
			this.store = store;
		}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
//ORIGINAL LINE: public void load() throws java.io.IOException
		public virtual void load()
		{
			ItemLoader loader = new ItemLoader();

			Storage storage = store.Storage;
			Index index = store.getIndex(IndexType.CONFIGS);
			Archive archive = index.getArchive(ConfigType.ITEM.Id);

			sbyte[] archiveData = storage.loadArchive(archive);
			ArchiveFiles files = archive.getFiles(archiveData);

			foreach (FSFile f in files.Files)
			{
				ItemDefinition def = loader.load(f.FileId, f.Contents);
				items[f.FileId] = def;
			}
		}

		public virtual ICollection<ItemDefinition> Items
		{
			get
			{
				// return Collections.unmodifiableCollection(items.Values);
				return new List<ItemDefinition>(items.Values);
			}
		}

		public virtual ItemDefinition getItem(int itemId)
		{
			return items[itemId];
		}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
//ORIGINAL LINE: public void export(java.io.File out) throws java.io.IOException
		public virtual void export(string @out)
		{
			// @out.mkdirs(); // TODO: ????

			foreach (ItemDefinition def in items.Values)
			{
				ItemExporter exporter = new ItemExporter(def);

				string targ = $"{@out}/{def.id}.json";
				exporter.exportTo(targ);
			}
		}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
//ORIGINAL LINE: public void java(java.io.File java) throws java.io.IOException
		public virtual void java(string java)
		{
			Console.WriteLine($"ItemManager.java not implemented! {java}");
			// java.mkdirs();
			// using (IDClass ids = IDClass.create(java, "ItemID"), IDClass nulls = IDClass.create(java, "NullItemID"))
			// {
			// 	foreach (ItemDefinition def in items.Values)
			// 	{
			// 		if (def.name.Equals("NULL", StringComparison.OrdinalIgnoreCase))
			// 		{
			// 			nulls.add(def.name, def.id);
			// 		}
			// 		else
			// 		{
			// 			ids.add(def.name, def.id);
			// 		}
			// 	}
			// }
		}

		public virtual ItemDefinition provide(int itemId)
		{
			return getItem(itemId);
		}
	}

}

File: src/main/java/net/runelite/cache/ItemManager.cpp
#include "ItemManager.h"
#include "IndexType.h"
#include "ConfigType.h"

namespace net::runelite::cache
{
	using File = java::io::File;
	using IOException = java::io::IOException;
	using Collection = java::util::Collection;
	using Collections = java::util::Collections;
	using HashMap = java::util::HashMap;
	using Map = java::util::Map;
	using ItemDefinition = net::runelite::cache::definitions::ItemDefinition;
	using ItemExporter = net::runelite::cache::definitions::exporters::ItemExporter;
	using ItemLoader = net::runelite::cache::definitions::loaders::ItemLoader;
	using ItemProvider = net::runelite::cache::definitions::providers::ItemProvider;
	using Archive = net::runelite::cache::fs::Archive;
	using ArchiveFiles = net::runelite::cache::fs::ArchiveFiles;
	using FSFile = net::runelite::cache::fs::FSFile;
	using Index = net::runelite::cache::fs::Index;
	using Storage = net::runelite::cache::fs::Storage;
	using Store = net::runelite::cache::fs::Store;
	using IDClass = net::runelite::cache::util::IDClass;

	ItemManager::ItemManager(std::shared_ptr<Store> store) : store(store)
	{
	}

	void ItemManager::load()
	{
		std::shared_ptr<ItemLoader> loader = std::make_shared<ItemLoader>();

		std::shared_ptr<Storage> storage = store->getStorage();
		std::shared_ptr<Index> index = store->getIndex(IndexType::CONFIGS);
		std::shared_ptr<Archive> archive = index->getArchive(ConfigType::ITEM.getId());

		std::vector<signed char> archiveData = storage->loadArchive(archive);
		std::shared_ptr<ArchiveFiles> files = archive->getFiles(archiveData);

		for (auto f : files->getFiles())
		{
			std::shared_ptr<ItemDefinition> def = loader->load(f->getFileId(), f->getContents());
			items.emplace(f->getFileId(), def);
		}
	}

	void ItemManager::link()
	{
		for (auto oc : items)
		{
			link(oc->second);
		}
	}

	void ItemManager::link(std::shared_ptr<ItemDefinition> item)
	{
		if (item->notedTemplate != -1)
		{
			item->linkNote(getItem(item->notedTemplate), getItem(item->notedID));
		}
		if (item->boughtTemplateId != -1)
		{
			item->linkBought(getItem(item->boughtTemplateId), getItem(item->boughtId));
		}
		if (item->placeholderTemplateId != -1)
		{
			item->linkPlaceholder(getItem(item->placeholderTemplateId), getItem(item->placeholderId));
		}
	}

	std::shared_ptr<Collection<std::shared_ptr<ItemDefinition>>> ItemManager::getItems()
	{
		return Collections::unmodifiableCollection(items.values());
	}

	std::shared_ptr<ItemDefinition> ItemManager::getItem(int itemId)
	{
		return items[itemId];
	}

	void ItemManager::export_Keyword(std::shared_ptr<File> out)
	{
		out->mkdirs();

		for (auto def : items)
		{
			std::shared_ptr<ItemExporter> exporter = std::make_shared<ItemExporter>(def->second);

			std::shared_ptr<File> targ = std::make_shared<File>(out, def->second.id + L".json");
			exporter->exportTo(targ);
		}
	}

	void ItemManager::java(std::shared_ptr<File> java)
	{
		java->mkdirs();
//JAVA TO C++ CONVERTER NOTE: The following 'try with resources' block is replaced by its C++ equivalent:
//ORIGINAL LINE: try (net.runelite.cache.util.IDClass ids = net.runelite.cache.util.IDClass.create(java, "ItemID"); net.runelite.cache.util.IDClass nulls = net.runelite.cache.util.IDClass.create(java, "NullItemID"))
		{
			IDClass ids = IDClass::create(java, L"ItemID");
			IDClass nulls = IDClass::create(java, L"NullItemID");
			for (auto def : items)
			{
				if (def->second.name.equalsIgnoreCase(L"NULL"))
				{
					nulls.add(def->second.name, def->second.id);
				}
				else
				{
					ids.add(def->second.name, def->second.id);
				}
			}
		}
	}

	std::shared_ptr<ItemDefinition> ItemManager::provide(int itemId)
	{
		return getItem(itemId);
	}
}
